File: src/0850/S0850_v2.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int rectangleArea(vector<vector<int>>& rectangles) {
        vector<vector<int>> ar(50);
        vector<vector<int>> tmp(50);
        tmp = rectangles;
        max_area(tmp, res);
        while (ar.size() != 0) {
            ar.clear();
            for (vector<int> j : tmp) {
                Solution::reOverlap(maxarea, j, ar);
            }
            max_area(ar, res);
            if (tmp == ar) break;
            tmp = ar;
        }
        if (ar.size() != 0) {
            for (vector<int> k : ar) {
                res += Solution::vect_area(k);
            }
        }
        return res % 1000000007;
    }

private:
    vector<int> maxarea;
    vector<int> vect;
    long long res = 0;

    void reOverlap(vector<int>& a, vector<int>& b, vector<vector<int>>& v) {
        if (a[0] <= b[0] && a[1] <= b[1] && a[2] >= b[2] && a[3] >= b[3]) { return; }
        else if (a[0] >= b[2] || a[1] >= b[3] || a[2] <= b[0] || a[3] <= b[1]) { v.push_back(b); }
        else if (a[0] <= b[0] && a[2] >= b[2]) {
            if (a[1] > b[1]) {
                if (a[3] < b[3]) { push_vec(v, b[0], a[3], b[2], b[3]); push_vec(v, b[0], b[1], b[2], a[1]); }
                else { if (a[1] < b[3]) { push_vec(v, b[0], b[1], b[2], a[1]); } }
            }
            else { if (a[3] > b[1] && a[3] < b[3]) { push_vec(v, b[0], a[3], b[2], b[3]); } }
        }
        else if (a[1] <= b[1] && a[3] >= b[3]) {
            if (a[0] > b[0]) {
                if (a[2] < b[2]) { push_vec(v, a[2], b[1], b[2], b[3]); push_vec(v, b[0], b[1], a[0], b[3]); }
                else { if (a[0] < b[2]) { push_vec(v, b[0], b[1], a[0], b[3]); } }
            }
            else { if (a[2]<b[2] && a[2]>b[0]) { push_vec(v, a[2], b[1], b[2], b[3]); } }
        }
        else if (a[0] > b[0]) {
            if (a[3] > b[3]) {
                if (a[2] >= b[2]) { push_vec(v, b[0], b[1], b[2], a[1]); push_vec(v, b[0], a[1], a[0], b[3]); }
                else { push_vec(v, b[0], b[1], b[2], a[1]); push_vec(v, b[0], a[1], a[0], b[3]); push_vec(v, a[2], a[1], b[2], b[3]); }
            }
            else if (a[1] < b[1]) {
                if (a[2] >= b[2]) { push_vec(v, b[0], a[3], b[2], b[3]); push_vec(v, b[0], b[1], a[0], a[3]); }
                else { push_vec(v, b[0], a[3], b[2], b[3]); push_vec(v, b[0], b[1], a[0], a[3]); push_vec(v, a[2], b[1], b[2], b[3]); }
            }
            else { push_vec(v, b[0], b[1], a[0], b[3]); push_vec(v, a[0], a[3], b[2], b[3]); push_vec(v, a[0], b[1], b[2], a[1]); }
        }
        else {
            if (a[2] < b[2]) {
                if (a[3] > b[3]) { push_vec(v, b[0], b[1], b[2], a[1]); push_vec(v, a[2], a[1], b[2], b[3]); }
                else if (a[1] < b[1]) { push_vec(v, b[0], a[3], b[2], b[3]); push_vec(v, a[2], b[1], b[2], a[3]); }
                else { push_vec(v, b[0], b[1], a[2], a[1]); push_vec(v, a[2], b[1], b[2], b[3]); push_vec(v, b[0], a[3], a[2], b[3]); }
            }
        }
    }

    inline void push_vec(vector<vector<int>>& arr, int& a, int& b, int& c, int& d) {
        vect = { a, b, c, d };
        arr.push_back(vect);
    }

    static inline long long vect_area(vector<int>& x) {
        static long long fa;
        static long long fb;
        fa = x[2];
        fa -= x[0];
        fb = x[3];
        fb -= x[1];
        fa *= fb;
        if (fa >= 0) { return fa; }
        else { return -fa; }
    }

    inline void max_area(vector<vector<int>>& x, long long& res) {
        maxarea = { 0, 0, 0, 0 };
        for (int i = 0; i < x.size(); i++) {
            maxarea = (vect_area(x[i]) > vect_area(maxarea)) ? x[i] : maxarea;
        }
        res += vect_area(maxarea);
    }
};

File: src/0850/S0850_v2.py
class Solution:
    def rectangleArea(self, rectangles) -> int:
        ar, tmp = [], []
        for i in rectangles:
            i.append(self.vectArea(i))
            tmp.append(tuple(i))
        res = 0
        while len(tmp) > 1:
            tmp.sort(key=lambda x: x[4], reverse=True)
            res += tmp[0][4]
            for j in tmp[1:]:
                self.reOverlap(tmp[0], j, ar)
            if tmp == ar:
                break
            tmp = []
            tmp, ar = ar, tmp
        if len(tmp) > 0:
            for k in tmp:
                res += k[4]
        if res == 0:
            res += tmp[0][4] 
        return res % 1000000007

    def vectArea(self, x):
        area = (x[2]-x[0])*(x[3]-x[1])
        return area if area >= 0 else -area

    def pushT(self, a, b, c, d, arr):
        arr.append((a, b, c, d, abs((c-a)*(d-b))))

    def reOverlap(self, a, b, v):
        if a[0] <= b[0] and a[1] <= b[1] and a[2] >= b[2] and a[3] >= b[3]:
            return
        elif a[0] >= b[2] or a[1] >= b[3] or a[2] <= b[0] or a[3] <= b[1]:
            v.append(b)
        elif a[0] <= b[0] and a[2] >= b[2]:
            if (a[1] > b[1]):
                if a[3] < b[3]:
                    self.pushT(b[0], a[3], b[2], b[3], v)
                    self.pushT(b[0], b[1], b[2], a[1], v)
                else:
                    if a[1] < b[3]:
                        self.pushT(b[0], b[1], b[2], a[1], v)
            elif a[3] > b[1] and a[3] < b[3]:
                self.pushT(b[0], a[3], b[2], b[3], v)
        elif a[1] <= b[1] and a[3] >= b[3]:
            if a[0] > b[0]:
                if a[2] < b[2]:
                    self.pushT(a[2], b[1], b[2], b[3], v)
                    self.pushT(b[0], b[1], a[0], b[3], v)
                elif a[0] < b[2]:
                    self.pushT(b[0], b[1], a[0], b[3], v)
            elif a[2] < b[2] and a[2] > b[0]:
                self.pushT(a[2], b[1], b[2], b[3], v)
        elif a[0] > b[0]:
            if a[3] > b[3]:
                if a[2] >= b[2]:
                    self.pushT(b[0], b[1], b[2], a[1], v)
                    self.pushT(b[0], a[1], a[0], b[3], v)
                else:
                    self.pushT(b[0], b[1], b[2], a[1], v)
                    self.pushT(b[0], a[1], a[0], b[3], v)
                    self.pushT(a[2], a[1], b[2], b[3], v)
            elif a[1] < b[1]:
                if (a[2] >= b[2]):
                    self.pushT(b[0], a[3], b[2], b[3], v)
                    self.pushT(b[0], b[1], a[0], a[3], v)
                else:
                    self.pushT(b[0], a[3], b[2], b[3], v)
                    self.pushT(b[0], b[1], a[0], a[3], v)
                    self.pushT(a[2], b[1], b[2], b[3], v)
            else:
                self.pushT(b[0], b[1], a[0], b[3], v)
                self.pushT(a[0], a[3], b[2], b[3], v)
                self.pushT(a[0], b[1], b[2], a[1], v)
        else:
            if a[2] < b[2]:
                if a[3] > b[3]:
                    self.pushT(b[0], b[1], b[2], a[1], v)
                    self.pushT(a[2], a[1], b[2], b[3], v)
                elif a[1] < b[1]:
                    self.pushT(b[0], a[3], b[2], b[3], v)
                    self.pushT(a[2], b[1], b[2], a[3], v)
                else:
                    self.pushT(b[0], b[1], a[2], a[1], v)
                    self.pushT(a[2], b[1], b[2], b[3], v)
                    self.pushT(b[0], a[3], a[2], b[3], v)
